### PR TITLE
Avoids exception when pandas.io.json.json_normalize contains items in…

### PR DESCRIPTION
**Pull request by dickreuter**
Continued in https://github.com/pandas-dev/pandas/pull/14583

---

When using pandas.io.json.json_normalize to parse a nested json and convert it to a dataframe, the meta parameter can be used to use fields as metadata for each record in resulting table. In some cases, not all items may contain all of the specified meta fields. This change will avoid throwing an error and output np.nan instead.

… meta parameter that don't always occur in every item of the list

### CODE REVIEW

**Comment by jreback** [comment]
what exactly is this fixing? would need a test.

**Comment by dickreuter** [comment]
{
	"Trades" : [{
			"general" : {
				"tradeid" : 100,
				"trade_version" : 1,
				"stocks" : [{

						"symbol" : "AAPL",
						"name" : "Apple",
						"price" : "0"

					}, {

						"symbol" : "GOOG",
						"name" : "Google",
						"price" : "0"

					}
				]
			},
		}, {
			"general" : {
				"tradeid" : 100,
				"stocks" : [{

						"symbol" : "AAPL",
						"name" : "Apple",
						"price" : "0"

					}, {
						"symbol" : "GOOG",
						"name" : "Google",
						"price" : "0"

					}
				]
			},
		}
	]
}

The following will fail because trade_version is only available in one of the two items in the list, so there is no way to output it if not all elements are exactly the same. With my change it will simply ignore it and output nan instead of throwing an error.

json_normalize(data=j['Trades'], record_path=[['general','stocks']], meta=[['general','tradeid'],['general','trade_version']])

**Comment by TomAugspurger** [comment]
@dickreuter Can you add that as a test and add a release note? See http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-to-the-code-base

**Comment by dickreuter** [comment]
This would be the test, but unclear where I should store it. Any suggestions?

from unittest import TestCase
from pandas.io.json import json_normalize

```

class TestDecision(TestCase):
    def test_json_normalise_fix(self):
        j = {
            "Trades": [{
                "general": {
                    "tradeid": 100,
                    "trade_version": 1,
                    "stocks": [{

                        "symbol": "AAPL",
                        "name": "Apple",
                        "price": "0"

                    }, {

                        "symbol": "GOOG",
                        "name": "Google",
                        "price": "0"

                    }
                    ]
                },
            }, {
                "general": {
                    "tradeid": 100,
                    "stocks": [{

                        "symbol": "AAPL",
                        "name": "Apple",
                        "price": "0"

                    }, {
                        "symbol": "GOOG",
                        "name": "Google",
                        "price": "0"

                    }
                    ]
                },
            }
            ]
        }
        j = json_normalize(data=j['Trades'], record_path=[['general', 'stocks']],
                           meta=[['general', 'tradeid'], ['general', 'trade_version']])
        self.assertEqual(len(j), 4)

```

**Comment by TomAugspurger** [comment]
Looks like those tests are all in https://github.com/pandas-dev/pandas/blob/master/pandas/io/tests/json/test_json_norm.py

You could add it as a test method under `TestJSONNormalize`

**Comment by dickreuter** [comment]
Test and documentation is now added.

File: doc/source/whatsnew/v0.19.1.txt [hunk 1]
```diff
@@ -78,3 +78,4 @@ Bug Fixes
 
 
 - Bug in ``pd.pivot_table`` may raise ``TypeError`` or ``ValueError`` when ``index`` or ``columns`` is not scalar and ``values`` is not specified (:issue:`14380`)
+- Bug in ``pandas.io.json.json_normalize``When parsing a nested json and convert it to a dataframe, the meta parameter can be used to use fields as metadata for each record in resulting table. In some cases, not all items may contain all of the specified meta fields. This change will avoid throwing an error and output np.nan instead. (:issue '14505')
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.1.txt]: pls simplify. a user just wants to know does this issue pertain to them, and a short expl.

make the issue

``(:issue:`14505`)``

File: pandas/io/json.py [hunk 2]
```diff
@@ -792,7 +792,10 @@ def _recursive_extract(data, path, seen_meta, level=0):
                     if level + 1 > len(val):
                         meta_val = seen_meta[key]
                     else:
-                        meta_val = _pull_field(obj, val[level:])
+                        try:
+                            meta_val = _pull_field(obj, val[level:])
+                        except:
```
Comment (jreback) [inline pandas/io/json.py]: don't use a bare except, list a specific exception ``KeyError``?

File: pandas/io/tests/json/test_json_norm.py [hunk 3]
```diff
@@ -225,6 +225,51 @@ def test_nested_flattens(self):
 
         self.assertEqual(result, expected)
 
+    def test_json_normalise_fix(self):
+        j = {
```
Comment (jreback) [inline pandas/io/tests/json/test_json_norm.py]: add the issue number as a comment

File: pandas/io/tests/json/test_json_norm.py [hunk 4]
```diff
@@ -225,6 +225,51 @@ def test_nested_flattens(self):
...
+
+                    }, {
+
+                        "symbol": "GOOG",
+                        "name": "Google",
+                        "price": "0"
+
+                    }
+                    ]
+                },
```
Comment (jreback) [inline pandas/io/tests/json/test_json_norm.py]: this prob does not pass linting. make sure it does.

File: pandas/io/tests/json/test_json_norm.py [hunk 5]
```diff
@@ -225,6 +225,51 @@ def test_nested_flattens(self):
...
+            }
+            ]
+        }
+        j = json_normalize(data=j['Trades'], record_path=[['general', 'stocks']],
+                           meta=[['general', 'tradeid'], ['general', 'trade_version']])
+        self.assertEqual(len(j), 4)
```
Comment (jreback) [inline pandas/io/tests/json/test_json_norm.py]: construct the expected frame and use ``assert_frame_equal``

File: pandas/io/json.py [hunk 6]
```diff
@@ -792,7 +792,10 @@ def _recursive_extract(data, path, seen_meta, level=0):
                     if level + 1 > len(val):
                         meta_val = seen_meta[key]
                     else:
-                        meta_val = _pull_field(obj, val[level:])
```
Comment (jreback) [inline pandas/io/json.py]: I think this should be a keyword, call it ``errors='raise'|'ignore'``. You are defining ``ignore``. Please leave the default as ``raise`` (which is the current behavior).

**Comment by dickreuter** [comment]
Added keyword errors {'raise'|'ignore}
Added documentation
Shortened what's new
Removed commas in dictionary for linting compatibility

**Comment by jreback** [comment]
you need to rebase on master

```
git rebase -i origin/master
```

**Comment by dickreuter** [comment]
I did that earlier today. It now says: "This branch is 8 commits ahead of pandas-dev:master."

**Comment by jreback** [comment]
maybe you didn't push it
it's not about conflicts

**Comment by dickreuter** [comment]
My fork on github seems up to date with what I have locally, so I assume it has been pushed. Are there any further changes you expected me to implement that are not present?

**Comment by jreback** [comment]
@dickreuter its impossible to see until you rebase on master. this should have just your commits

https://github.com/pandas-dev/pandas/pull/14505/commits

**Comment by dickreuter** [comment]
I see, isn't it showing those commits of others only because I did a rebase of my fork (and then a local rebase of my local copy instead of a merge?). If that's a problem I could delete my fork and create a new one, then make the changes again and create a new pull request, unless you have a better suggestion.

**Comment by jreback** [comment]
you prob just need something like

```
git fetch origin
git rebase -i origin/master
git push yourremote thisbranchname -f
```

**Comment by dickreuter** [comment]
git fetch origin --> doesn't fetch anything as my local copy is in
sync with my remote origingit rebase -i origin/master --> noop - no
commits to pick

I think all my changes can be seen
here:https://github.com/pandas-dev/pandas/pull/14505/commits/984883796e0ee6e7355c2d5a746273c9551ead91

What may be confusing is that I also did an automatic reformatting
with pycharm to make the file conform with pep8. But those changes
only concern empty spaces.

**Comment by jreback** [comment]
could also be called ``upstream``
```
(pandas) [Thu Nov 03 19:52:46 ~/pandas]$ git remote -v|grep origin
origin  https://github.com/pandas-dev/pandas.git (fetch)
origin  https://github.com/pandas-dev/pandas.git (push)
```

it doesn't matter if your branch is in sync with YOUR upstream, rather it needs to be in sync with pandas master (and on top of it), that's what a rebase is.

you need to rebase to remove all of the merges of master. you shouldn't do that, instead rebase.

**Comment by dickreuter** [comment]
This seems to be the problem:
http://stackoverflow.com/questions/40413071/after-rebasing-my-github-fork-commits-from-others-are-in-my-pull-request/40413455#40413455

Will try to fix it, or if it's too complicated just delete and redo.

**Comment by jorisvandenbossche** [comment]
@jreback Regarding your comment above (https://github.com/pandas-dev/pandas/pull/14505#issuecomment-258305113): pandas-dev repo is typically called 'upstream', and your own fork 'origin' (that's how our contributor guide also says it), so you need `git fetch upstream` and `git rebase -i upstream/master` instead of `git fetch origin` and `git rebase -i origin/master` to rebase properly.

**Comment by jorisvandenbossche** [comment]
Follow-up in #14583

**Pull request closed without merging**